### ISSUE TITLE
Embedding layer with mask_zero=True gives unexpected weighted accuracy

### ISSUE THREAD

**Issue by gustavgransbo** [keras-team/keras#10267]
When evaluating a model using weighted accuracy in combination with setting `mask_zero=True` in my Embedding layer I encountered unexpectedly high weighted accuracy. 

To me it seems like weighted accuracy should not be affected by setting mask_zero=True. Though, I also assume weighted accuracy is not necessary when `mask_zero=True`, so this might not be a very important issue.

Anyway, here is a toy example to reproduce the problem:
```python
import numpy as np
from keras.layers import Embedding, Input
from keras.models import Model

# Set up the toy problem
# Input consists of variable length sequences of 1's and 2's padded with 0's
# Targets are one-hot encoded 1's for all instances, 0's  for the padding
X = [
    [1, 1, 1, 0],
    [1, 1, 2, 2],
    [1, 0, 0, 0],
    [2, 2, 0, 0]
]
X = np.asarray(X)
y = np.asarray([[[0, 1, 0] if x else [1, 0, 0] for x in instance] for instance in X])

# Initialize sample weight, 1 for all training instances and 0 for the padding
sample_weights = (X != 0).astype(int)

# Set up the embedding layer. It embeds the inputs (including the padding) with one-hot encoding
n_words = 2
dims = n_words+1
embedding_matrix = np.eye(dims)
embedding = Embedding(n_words+1, dims, weights=[embedding_matrix], trainable=False, mask_zero=True)

# Set up the model. It simply embeds the input and outputs it
model_input = Input(shape=(None,))
model_output = embedding(model_input)
model = Model(model_input, model_output)

# Compile and evaluate the model
# The model should successfully classify all 1's, fail on the 2's and ignore the padding
# This should, and does result in an accuracy of 60%
# However, it results in an unexpected weighted_accuracy of 96%
model.compile(optimizer='adam', loss='categorical_crossentropy', weighted_metrics=['acc'], metrics=['acc'], sample_weight_mode='temporal')

print("Loss: {}, Accuracy {}, Weigted Accuracy: {}".format(*model.evaluate(X, y, sample_weight=sample_weights)))
```
Output:
`Loss: 10.315580368041992, Accuracy 0.6000000238418579, Weigted Accuracy: 0.9599999785423279`

**Comment by maym2104** [comment]
I have a model trained on sequential data but cannot use the embedding layer. So I only use weighted_loss et weighted_accuracy as you do by having sample weights that are basically a mask on some output. So I cannot compare with the accuracy value. 
Recently I change my training sample weights to give more weight on examples later in the sequences, because there was much less long sequences than short sequences in my dataset and thus thought I was overfitting short sequences, which had much higher accuracy. But now, I have > 1 weighted accuracy, ranging from 2 to 9. So there seems to be something fishy going on when computing weighting_accuracy.

**Comment by maym2104** [comment]
oh wow way to go Keras. Just looked at code, and basically they do not divide by sum of all weights, but by a ratio of number of weights that are different to 0. So in my case I have weights different from 0 but also 1, I dunno what I get but it is totally not the weighted mean of my loss nor my accuracy.
@guohengkai In your case though the problem might be that the weighted accuracy already takes into account the mask of your data. So you compute the weighted accuracy of your already weighted accuracy. Look at the code at https://github.com/keras-team/keras/blob/master/keras/engine/training_utils.py#L415.
